src/Features/CSharp/Portable/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,105 @@
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeNode,
+                SyntaxKind.ForStatement,
+                SyntaxKind.EqualsValueClause,
+                SyntaxKind.IfStatement,
+                SyntaxKind.WhenClause,
+                SyntaxKind.WhileStatement,
+                SyntaxKind.DoStatement,
+                SyntaxKind.ReturnStatement,
+                SyntaxKind.YieldReturnStatement,
+                SyntaxKind.SimpleAssignmentExpression,
+                SyntaxKind.ArrowExpressionClause,
+                SyntaxKind.SimpleLambdaExpression,
+                SyntaxKind.ParenthesizedLambdaExpression,
+                SyntaxKind.Argument);
+
+        private static ExpressionSyntax? GetExpression(SyntaxNode node)
+            => node switch
+            {
+                ForStatementSyntax n => n.Condition,
+                EqualsValueClauseSyntax n => n.Value,
+                IfStatementSyntax n => n.Condition,
+                WhenClauseSyntax n => n.Condition,
+                WhileStatementSyntax n => n.Condition,
+                DoStatementSyntax n => n.Condition,
+                ReturnStatementSyntax n => n.Expression,
+                YieldStatementSyntax n => n.Expression,
+                ArrowExpressionClauseSyntax n => n.Expression,
+                AssignmentExpressionSyntax n => n.Right,
+                LambdaExpressionSyntax n => n.ExpressionBody,
+                ArgumentSyntax n when n.GetRefKind() == RefKind.None => n.Expression,
+                _ => null,
+            };
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            // TODO need an option for user to disable the feature
+
+            // TODO need to check language version >= C# 9.0
+
+            var parentNode = context.Node;
+
+            var expression = GetExpression(parentNode);
+            if (expression is null)
+                return;
+
+            var operation = context.SemanticModel.GetOperation(expression, context.CancellationToken);
+            if (operation is null)
+                return;
+
+            if (!CSharpUsePatternCombinatorsAnalyzer.Analyze(operation, out var pattern, target: out _))
+                return;
+
+            if (IsTrivial(pattern))
+                return;
+
+            context.ReportDiagnostic(DiagnosticHelper.Create(
+                Descriptor,
+                location: expression.GetLocation(),
+                effectiveSeverity: ReportDiagnostic.Warn,
+                additionalLocations: null,
+                properties: null,
+                messageArgs: null));
+        }
+
+        private static bool IsTrivial(AnalyzedPattern pattern)
+            => pattern switch
+            {
+                Not { Pattern: Constant _ } => true,
+                Not _ => false,
+                Binary _ => false,
```

Comment:
interessting.  why?   what is the intuition for when we would want to show/not-show this?

Reply:
this is 'use pattern combinators' so we do just that. 
the code can generate `if (is null)`, `if (is not null)`, `if (is > 0)` if you don't stop it.